ts/src/test/test.ts

```diff
@@ -252,30 +249,63 @@ export default class testMainClass extends baseMainTestClass {
+                    dump ('[TEST_WARNING]', 'Authentication problem for public method', exceptionMessage (e), exchange.id, methodNameInTest, argsStringified);
+                }
+            } else {
                 throw e;
             }
         }
-        return result;
     }
 
     async testSafe (methodName, exchange, args, isPublic) {
-        try {
-            await this.testMethod (methodName, exchange, args, isPublic);
-            return true;
-        } catch (e) {
-            return false;
+        // `testSafe` method does not throw an exception, instead mutes it.
+        // The reason we mute the thrown exceptions here is because if this test is part
+        // of "runPublicTests", then we don't want to stop the whole test if any single
+        // test-method fails. For example, if "fetchOrderBook" public test fails, we still
+        // want to run "fetchTickers" and other methods. However, independently this fact,
+        // from those test-methods we still echo-out (console.log/print...) the exception
+        // messages with specific formatted message "[TEST_FAILURE] ..." and that output is
+        // then regex-parsed by run-tests.js, so the exceptions are still printed out to
+        // console from there. So, even if some public tests fail, the script will continue
+        // doing other things (testing other spot/swap or private tests ...)
+        const maxRetries = 5;
+        const argsStringified = '(' + args.join (',') + ')';
+        for (let i = 0; i < maxRetries; i++) {
+            try {
+                await this.testMethod (methodName, exchange, args, isPublic);
+                return true;
+            } catch (e) {
+                const isRateLimitExceeded = (e instanceof RateLimitExceeded);
+                const isExchangeNotAvailable = (e instanceof ExchangeNotAvailable);
+                const isNetworkError = (e instanceof NetworkError);
+                const isDDoSProtection = (e instanceof DDoSProtection);
+                const isRequestTimeout = (e instanceof RequestTimeout);
+                const tempFailure = (isRateLimitExceeded || isExchangeNotAvailable || isNetworkError || isDDoSProtection || isRequestTimeout);
+                if (tempFailure) {
+                    // wait and retry again
+                    await exchange.sleep (i * 1000); // increase wait seconds on every retry
+                    continue;
+                } else {
+                    // if not temp failure, then dump exception without retrying
+                    dump ('[TEST_WARNING]', 'Method could not be tested', exceptionMessage (e), exchange.id, methodName, argsStringified);
+                    return false;
+                }
+            }
         }
+        // if maxretries was gone with same `tempFailure` error, then let's eventually return false
+        dump ('[TEST_WARNING]', 'Method could not be tested, because of temporary access issues', exchange.id, methodName, argsStringified);
```

Comment:
@ttodua  this message is a bit strange, can we replace with something like "Method not tested due to a Network/Availability issue

Reply:
one note I can say is that it can be one of the exceptions: RL, Network, Timeout, etc... so the term might be somewhat in-between of those.